mozfldp/fed_avg_w_dp.py

```diff
@@ -0,0 +1,250 @@
+    prev_theta = np.array(theta_0, copy=True)
+    user_sel_prob = 1.0 / prms.num_users
+    standard_dev = _calc_standard_dev(
+        prms.noise_scale, prms.sensitivity, user_sel_prob, weight_sum
+    )  # This feels weird being a constant...
+    user_updates_buf = []
+
+    for round_t in range(prms.num_rounds):
+        # Pick a random set of users to sample
+        random_user_idxs_sample = _get_random_selection_of_user_idxs(prms.num_users)
+
+        # Query the selected users
+        user_updates_buf.clear()
+        for user_idx in random_user_idxs_sample:
+            user_round_labels, user_round_feats = _get_data_for_user_for_round(
+                prms, data, user_idx
+            )
+            user_updates_buf.append(
+                user_update_fed_avg(prms, user_round_feats, user_round_labels, theta_0)
+            )
+
+        # Merge (fc)
+        merged_user_values = _merge_all_user_weights(
+            user_sel_prob, weight_sum, user_updates_buf, user_weights
+        )  # Note that we start at t + 1
+
+        # Note: Assuming for now that S is defined before we run
+        rand_gauss_noise = _gen_gausian_rand_noise(
+            standard_dev, len(merged_user_values)
+        )
+
+        theta = prev_theta + merged_user_values + rand_gauss_noise
+        prev_theta = theta
+
+        print("Theta ([coef, inter] for round {}: \n {}".format(round_t, theta))
+        _moments_accountant_accum_priv_spending(prms.noise_scale)
+
+    privacy_spent = _calc_privacy_spent()
+    print("Total privacy spent from privacy budget: {}".format(privacy_spent))
+
+    coef_slice, inter_slice = _get_coef_and_inter_slice_from_theta(
+        theta, prms.num_features, prms.num_labels
+    )
+
+    return np.array(coef_slice), np.array(inter_slice)
+
+
+# TODO: Give better function name...
+def _merge_all_user_weights(user_sel_prob, weight_sum, user_updates_buf, user_weights):
+    """
+    Merge all user updates for a round into a single delta (vector).
+    """
+
+    num_users_in_batch = len(user_updates_buf)
+    merged_weights = np.zeros(len(user_updates_buf[0]), dtype=np.float64, order="C")
+
+    for i in range(num_users_in_batch):
+        weighted_user_val = np.multiply(user_updates_buf[i], user_weights[i])
+        merged_weights = np.add(merged_weights, weighted_user_val)
+
+    merged_weights = np.divide(merged_weights, user_sel_prob * weight_sum)
+    return merged_weights
+
+
+def flat_clip(sensitivity, vec):
+    vec *= min(1, sensitivity / np.linalg.norm(vec))
+    return vec
+
+
+def user_update_fed_avg(prms, round_user_features, round_user_labels, theta_0):
+
+    batches_features = []
+    batches_labels = []
+
+    classifier = SGDClassifier(
+        loss="hinge", penalty="l2", max_iter=1, random_state=prms.rand_seed
+    )
+
+    round_num_entries = len(round_user_features)
+
+    # For now just skip any batches that are smaller than batch_size
+    num_batches = round_num_entries // prms.batch_size
```

Comment:
I don't think there's any reason to drop the last batch, even if it's smaller than the target size.

Reply:
Agreed. Fixed.